Added Chamberlain MyQ Binding

computergeek1507 opened the pull request:
@teichsta Chamberlain MyQ Binding is ready for review

watou commented on bundles/binding/org.openhab.binding.myq/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,44 @@
+Bundle-Name: openHAB myq Binding
+Bundle-SymbolicName: org.openhab.binding.myq
+Bundle-Vendor: openHAB.org
+Bundle-Version: 1.0.0.qualifier
```
Change to 1.8.0.qualifier

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/MyqBindingProvider.java:
```diff
@@ -0,0 +1,41 @@
+	 * @return List of item names mapped to a Chamberlain myQ Binding
+	 */
+	public List<String> getInBindingItemNames();
+}
```
please add newline

computergeek1507 replied:
@watou sorry I do not understand where you want the newline, at the end of the file?

watou replied:
Yes, at the end of the file, thanks!

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/GarageDoorData.java:
```diff
@@ -0,0 +1,89 @@
+
+					if (deviceType.contains("Garage")
+							&& deviceType.contains("Door")
+							&& deviceType.contains("Opener")) {
```
Suggest instead using a static final java.util.regex.Pattern to match these three strings in case they change later.

digitaldan replied:
I'll get this one

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/GarageDoorDevice.java:
```diff
@@ -0,0 +1,127 @@
+	public String toString() {
+		return this.DeviceName;
+	}
+}
```
please add newline

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/LoginData.java:
```diff
@@ -0,0 +1,58 @@
+		Map<String, Object> treeData = mapper.readValue(root, Map.class);
+		Object data = treeData.get("SecurityToken");
+		if(data == null)
+			throw new IOException("Could not find SecurityToken in JSON data");
```
Create a new subclass of Exception instead of using IOException here, since it's not an I/O issue.

digitaldan replied:
I'll get this one

digitaldan replied:
@watou I could make a custom exception for this, but nothing needs to do anything different if the response is malformed VS something else that throws an IOException deeper down the network stack.  By adding a new exception type I would be requiring other parts of the code to catch additional exception types that they don't care about.  This seemed to be the cleanest way to me as I considered it an data input problem, is there something else that could work?

watou replied:
I think it would improve readability to formally use an exception type that represents the class of exception this is, which could be about malformed JSON.  I worry that higher catch blocks could take an IOException at face value that there are network issues, when in fact the I/O is fine but the JSON was not as expected.  For example, if a Java library class unexpectedly threw an IOException when there way no I/O problem, a catch block could potentially tear down and reconnect a socket that was not the problem at all.  Can higher catch blocks catch any shared parent class like Exception?

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/LoginData.java:
```diff
@@ -0,0 +1,58 @@
+	public String getSecurityToken() {
+		return securityToken;
+	}
+}
```
please add newline

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyQResponseCode.java:
```diff
@@ -0,0 +1,66 @@
+	  * @param code
+	  * @return
+	  */
+	 public static MyQResponseCode FromCode(int code){
```
Please use `fromCode` method naming style.

digitaldan replied:
I can grab this one

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBinding.java:
```diff
@@ -0,0 +1,369 @@
+ */
+public class MyqBinding extends AbstractBinding<MyqBindingProvider> {
+	private static final Logger logger = LoggerFactory
+			.getLogger(MyqBinding.class);
```
Please the openHAB Eclipse formatter over all Java files (120 max line length, etc.)

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBinding.java:
```diff
@@ -0,0 +1,369 @@
+	 * set in the activate() method and must not be accessed anymore once the
+	 * deactivate() method was called or before activate() was called.
+	 */
+	@SuppressWarnings("unused")
```
Is this still needed?

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBinding.java:
```diff
@@ -0,0 +1,369 @@
+	 */
+	private void poll() {
+		if (invalidCredentials || this.myqOnlineData == null) {
+			return;
```
Might be a good spot for a `logger.debug` message.

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBindingConfig.java:
```diff
@@ -0,0 +1,46 @@
+
+	int deviceIndex;
+	ITEMTYPE type;
+}
```
please add newline

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBindingConfig.java:
```diff
@@ -0,0 +1,46 @@
+	};
+
+	int deviceIndex;
+	ITEMTYPE type;
```
Instead of an `ITEMTYPE` enum, just keep the `Class<? extends Item>` of the item, so you're not creating an enum that partially mirrors the various item types in openHAB.

watou replied:
...or possibly better, save a copy of `item.getAcceptedDataTypes()`, and when it's time to post an update, loop through the list until you find a type you support.  That way you're independent of which kind of item it is you're posting to.

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqData.java:
```diff
@@ -0,0 +1,222 @@
+	 *            MyQ deviceID of Garage Door Opener.
+	 * @param state
+	 *            Desired state to put the door in, 1 = open, 0 = closed
+	 * @param attemps
```
please correct

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqData.java:
```diff
@@ -0,0 +1,222 @@
+			super(message);
+		}
+	}
+}
```
please add newline

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqData.java:
```diff
@@ -0,0 +1,222 @@
+	}
+
+	@SuppressWarnings("serial")
+	public class InvalidLoginException extends Exception {
```
should this be a static nested class instead of an inner class?

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqGenericBindingProvider.java:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	/**
+	 * @{inheritDoc
```
should be `{@inheritDoc}`

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqGenericBindingProvider.java:
```diff
@@ -0,0 +1,110 @@
+			config.type = ITEMTYPE.ContactStatus;
+		} else if (item instanceof StringItem) {
+			config.type = ITEMTYPE.StringStatus;
+		}
```
As mentioned before, instead of `ITEMTYPE`, consider
```
config.acceptedDataTypes = new ArrayList<Class<? extends State>>(item.getAcceptedDataTypes());
```
and stay unaware of the different item classes altogether.  Just post an update by looping the array and finding a type you support.

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqGenericBindingProvider.java:
```diff
@@ -0,0 +1,110 @@
+		}
+		return inBindings;
+	}
+}
```
please add newline

watou commented on distribution/openhabhome/configurations/openhab_default.cfg:
```diff
@@ -2180,3 +2180,12 @@ tcp:refreshinterval=250
+
+# Chamberlain MyQ Username and Password
+# myq:username=
+# myq:password=
```
please add newline

computergeek1507 commented:
@watou I think @digitaldan and I have integrated most comments you suggested. Please let me know if we missed something. Thanks.

watou commented:
Thanks very much for the binding.  I know lots of people use/used the Vera plugin to talk to the myq garage door opener, so this should be pretty popular!

I would like to take one more look in the next day or two, and thanks for the changes!

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/GarageDoorData.java:
```diff
@@ -0,0 +1,81 @@
+					String deviceType = node.get(i).get("MyQDeviceTypeName")
+							.asText();
+
+					if (deviceType.matches("Garage\\s?Door\\s?Opener")) {
```
This is expensive compared to having
```
private static final Pattern DEVICE_TYPE_PATTERN = Pattern.compile("Garage\\s?Door\\s?Opener");
//...
if (DEVICE_TYPE_PATTERN.matcher(deviceType).matches()) { 
//...
```

digitaldan replied:
yep

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBindingConfig.java:
```diff
@@ -0,0 +1,30 @@
+public class MyqBindingConfig implements BindingConfig {
+
+	int deviceIndex;
+	Item type;
```
Bindings should avoid holding onto `Item` objects because their lifecycle is independent of bindings.  I would replace this member with
```
List<Class<? extends State>> acceptedDataTypes;
```

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqGenericBindingProvider.java:
```diff
@@ -0,0 +1,101 @@
+			throws BindingConfigParseException {
+		final MyqBindingConfig config = new MyqBindingConfig();
+
+		config.type = item;		
```
Recommend
```
config.acceptedDataTypes = new ArrayList<Class<? extends State>>(item.getAcceptedDataTypes());
```

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBinding.java:
```diff
@@ -0,0 +1,374 @@
+					if (deviceConfig != null) {
+						GarageDoorDevice garageopener = garageStatus
+								.getDevice(deviceConfig.deviceIndex);
+						if (garageopener != null) {
```
Recommend
```
State newState = UnDefType.UNDEF;
for (Class<? extends State> type : deviceConfig.acceptedDataTypes) {
  if (StringType.class == type) {
    // newState = new StringType...
    break;
  }
  else if (OpenClosedType.class == type) {
    // newState = OpenClosedType...
    break;
  }
  else {
    // etc...
  }
}
eventPublisher.postUpdate(mygItemName, newState);
```
this way you don't care about the kind of item, only the state types it supports.

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqData.java:
```diff
@@ -0,0 +1,207 @@
+
+	private Properties header;
+	private final String webSite = "https://myqexternal.myqdevice.com";
+	private final String appId = "JVM/G9Nwih5BwKgNCjLxiFUQxQijAebyyg8QUHr7JOrP+tuPb8iHfRHKwTmDzHOu";
```
Is there any risk that some other API user could use and abuse this appId, as opposed to having the binding user register his/her own app and add it to configuration in openhab.cfg?

digitaldan replied:
Unfortunately we have to use an existing app id that's out in the wild.  This is also what the smart things plugin is doing and I'm assuming the vera plugin as well.  Not ideal, I know.

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqData.java:
```diff
@@ -0,0 +1,207 @@
+
+	private String userName;
+	private String password;
+	private String sercurityTokin;
```
`securityToken` possibly? :)

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqData.java:
```diff
@@ -0,0 +1,207 @@
+
+		logger.debug("Received MyQ  JSON: {}", dataString);
+
+		if (dataString == null)
```
Always use `{` and `}` even for one-liners.

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqGenericBindingProvider.java:
```diff
@@ -0,0 +1,101 @@
+	@Override
+	public void processBindingConfiguration(String context, Item item,
+			String bindingConfig) throws BindingConfigParseException {
+		super.processBindingConfiguration(context, item, bindingConfig);
```
move to end of method in case your code throws a `BindingConfigParseException`, in which case we don't want to pollute the internal openHAB data structures with the bad data.

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBinding.java:
```diff
@@ -0,0 +1,374 @@
+				command, itemName);
+		if (myqOnlineData != null) {
+			computeCommandForItem(command, itemName);
+		}
```
Might want to log a WARN if you are going to drop the command due to no `myqOnlineData`.

watou commented:
Please add to `distribution/pom.xml`.  I'm going to tag this PR awaiting-feedback, but if other maintainers would add review comments, please untag it while reviewing and retag after.  Thanks!

digitaldan commented:
@watou thanks for spending the time to thoroughly review the binding.  @computergeek1507 I have a PR around the regex pattern, I'll leave the rest up to you so we don't step on each others toes ;-)   If you want me to fix anything else please let me know!

computergeek1507 commented:
@watou I hope my latest commit is what you were thinking with the "acceptedDataTypes" stuff. @digitaldan Can you verify this new code still works with your setup. It works with my server but I just want another set of eyes to verify my code isn't crazy. Thanks everyone for the help.

digitaldan commented:
@computergeek1507 I just pulled and compiled your latest changes and all looks good, I'm out of town and its late here so I didn't actually open or close my door, but the door status is updating fine.

computergeek1507 commented:
@digitaldan Thanks for the look. I only really changed the status code so as long as it working I think it's good.

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/GarageDoorDevice.java:
```diff
@@ -0,0 +1,127 @@
+	private int DeviceId;
+	private String DeviceType;
+	private String DeviceName;
+	private GarageDoorStatus Status;
```
variable names should not begin with upper-case letters

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/InvalidDataException.java:
```diff
@@ -0,0 +1,17 @@
+ * @author Dan Cunningham
+ *
+ */
+public class InvalidDataException extends IOException {
```
Is `InvalidDataException` conceptually a subclass of `IOException`?  Since the transport guarantees that delivered data is not corrupted due to network problems, it seems to me that invalid data could not be the result of an I/O problem.

digitaldan replied:
This is how the Jackson json parser implements its exceptions ( like JsonParseException) if the JSON is malformed or unexpected. I liked this pattern and thought it fit this use case well as the developer can be very specific about the exception they want to catch or not.

watou replied:
sounds good, Dan -- thanks.

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBinding.java:
```diff
@@ -0,0 +1,381 @@
+									}
+								}
+							}
+							if (newState != UnDefType.UNDEF) {
```
Minor point: Why not post the UNDEF update?  That would give feedback to the user that the binding is explicitly reporting "I don't know how to update the item you've bound to me."

computergeek1507 replied:
@watou My intention is to use a switch item as a "push button" and trigger opening or closing event. Since I don't want the switch state to update I didn't add it to the if else statements and it will always be UNDEF.

watou replied:
Why not have a Switch item track an ON or OFF state, where you give the meaning that ON means open and OFF means closed (or the reverse if you prefer).  I think it would be of benefit to be able to use a Switch item for both state and commands, and not require an additional Contact item just to be able to do both state and commands.

The CMD servlet and REST API support the TOGGLE function, the UI can give you a toggle using visibility= for two switches in the sitemap, and rules can inspect the current Switch state in order to send the opposite ON or OFF command.  So having a "no-state" switch is not necessary.  Do you disagree?

computergeek1507 replied:
I didn't know what state to put the switch in when the garage is partially open but I could just say its on for partially open.

watou replied:
Since that is the same logic as for OpenClosedType and it makes perfect sense, I agree! :-)

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBinding.java:
```diff
@@ -0,0 +1,381 @@
+									}
+								} else if (PercentType.class == type) {
+									if (garageopener.getStatus() != GarageDoorStatus.UNKNOWN) {
+										newState = new PercentType(50);
```
Just curious: How does a 50% update derive from the garage door status?

digitaldan replied:
We know when the door is in motion or is stuck in in a half open state, we don't know exactly what percentage the door is open or closed. Roller shutter items afaik only support OPEN/CLOSE and percentage.  I put this there so that roller shutter items could display the correct state icon, so in my sitemap the  icon goes from close -> halfway open -> open.

watou replied:
As the `RollershutterItem` is currently defined
```
		acceptedDataTypes.add(UnDefType.class);
		acceptedDataTypes.add(UpDownType.class);
		acceptedDataTypes.add(PercentType.class);
```
the `UpDownType` appears earlier in the list of accepted data types, so I don't think that code will be hit, at least not for Rollershutter items.

watou replied:
Never mind -- I see now that the UpDownType check will be skipped and match the PercentType test on the next loop.

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqData.java:
```diff
@@ -0,0 +1,208 @@
+
+	private Properties header;
+	private final String webSite = "https://myqexternal.myqdevice.com";
+	private final String appId = "JVM/G9Nwih5BwKgNCjLxiFUQxQijAebyyg8QUHr7JOrP+tuPb8iHfRHKwTmDzHOu";
```
Is this safe to hard-code an app ID in the binding?  What external dependencies are implied by tying the binding code to this particular app ID?

digitaldan replied:
see my comment at the end

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqData.java:
```diff
@@ -0,0 +1,208 @@
+
+		String dataString = executeUrl(method, url, header,
+				payload == null ? null : IOUtils.toInputStream(payload),
+				payloadType, 5000);
```
Is there any chance at all that a 5-second timeout will in some case be insufficient?  If so, could this be a binding config value in openhab.cfg / myq.cfg ?

watou commented:
Thanks for the updates.  I've added some final comments and questions.  Among the others, I'm concerned about the hard-coded `appID` and who controls it, and if it would not be better to document in the wiki how the user is empowered to create and use their own appID via openhab.cfg / myq.cfg.  

If other maintainers want to chime in, please do.

digitaldan commented:
@watou thanks, regarding the app Id... I have gone back and forth about this.  The API only allows clients with the correct app id to access the API, users can not make up there own.  Furthermore Chamberlain seems to occasionally stop known Id's from working.  I feel like we are kinda walking on egg shells with them, and we are mindful of raising attention if they start getting requests from us with random app ids.  On the other hand, if they do depreciate an id I could see it being nice for users to update the binding on their own until we can release a updated version, this would be an advanced option and I would only want users to use this temporarily.  But on the other other hand, if they have an old ID in their config but get a new version of the binding with an updated id, then we start getting into a bad state.   

That was a very long way of me saying I think we should stick with it in the binding, I could be persuaded to make it an optional config override, but I worry over time that could actually hurt users when newer version of the binding come out that require an even newer id.

watou commented:
Hi Dan, at the very least I think users should be able to supply their own app ID in the config file, and not rely on one that is registered to and controlled by a third party person or company.  I also think that whatever steps you or Scott went through to obtain an app ID ought to be well documented in the wiki page (see the [Ecobee](https://github.com/openhab/openhab/wiki/Ecobee-Binding) and [Nest](https://github.com/openhab/openhab/wiki/Nest-Binding#binding-configuration) binding wiki pages by example).

> I worry over time that could actually hurt users when newer version of the binding come out that require an even newer id.

What code dependencies exist between an app ID and later versions of the binding?

Have you or Scott written a wiki page for the binding yet?

Thanks again for all the good work!

digitaldan commented:
@watou that's part of the problem, the only way to obtain an app id has been by sniffing traffic from android or IOS clients ;-(  Sad I know.  Chamberlain is very, well....,  unfriendly with the developer community.  I had to set up a reverse ssl proxy for my phone to get a valid ID.  As far as I know there may only be a half dozen or less (just a guess) working app id's. I feel strongly we should control the app id in the binding and not put the burden on the end user to figure out which ids work.  Also the last time ( 2 weeks ago) they expired the ID we were using they also modified their api to reject requests which were missing some seemingly unimportant http headers, in this case being able to change the ID would not have helped as it also signaled a breaking change in their API. 

If you still feel strongly about it, it's not a lot of work to put in, I just worry that if a user needs to change the id, more will be needed anyway to make the binding work. And if they do change it, updating the binding which depends on a newer id may then break it for the users who over wrote the id in their config (and will need to go back and comment their change out).

digitaldan commented:
@watou  There is a wiki!  @computergeek1507  has a page here https://github.com/computergeek1507/openhab/wiki/Chamberlain-MyQ-Binding  which he has been keepig up to date.

watou commented:
@digitaldan Now I understand where you're coming from, and why this could be a moving target.  Here is my thinking about it:

1. Absolutely put an explanation similar to what you just provided above in the wiki.  Users should know that the vendor could break the binding without notice, and possibly in such a way that it can't be quickly repaired and code changes might be required.
1. Make the appID configurable from the .cfg file (or if configuration is missing, try to use the one in the code) because there is a chance that a new app ID would make the binding work again, thus avoiding the need for a code change, and the vendor may yet decide to work with 3rd party developers and expect app IDs to be kept secure by developers (or in openHAB's case, to have individual users get their own app IDs).
1. When the vendor invalidates an app ID, how are they not breaking their Android and iOS clients in the process?  Do the app clients simply fail and need an update from the app store, or do they retrieve an app ID on a secure channel, or have a time-limited list of app IDs embedded in the apps, or what exactly?  Is it possible to work similarly to the apps, assuming they work better than simply failing?

digitaldan commented:
@watou 
1. I will update the wiki with a summary of the appid
2.  I have a PR open on @computergeek1507 repo to address the appId and timeout overrides in the config file.
3. I'm assuming they force users to update the app if the id gets retired.  To tell you the truth, I only started working on this 3 weeks ago so I am going on what I have read from reports on github projects and also over at the smarthings community ( which also has a community supported plugin).  There is not a lot of transparency on how the appid works.

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBinding.java:
```diff
@@ -0,0 +1,393 @@
+			GarageDoorData garageStatus = myqOnlineData.getGarageData();
+			GarageDoorDevice garageopener = garageStatus
+					.getDevice(deviceConfig.deviceIndex);
+			if (garageopener != null) {
```
As mentioned earlier, recommend against implementing toggle logic in binding.  So briefer code:
```
if (command.equals(OnOffType.ON) || command.equals(UpDownType.UP)) {
    myqOnlineData.executeGarageDoorCommand(garageopener.getDeviceId(), 1);
    beginRapidPoll(true);
} else if (command.equals(OnOffType.OFF) || command.equals(UpDownType.DOWN)) {
    myqOnlineData.executeGarageDoorCommand(garageopener.getDeviceId(), 0);
    beginRapidPoll(true);
} else {
    logger.warn("Unknown command {}", command);
}
```

computergeek1507 replied:
switch items have a toggle command and I don't know what type that will be.

computergeek1507 commented:
@watou I hope I have changed everything you have made recommendations on. @digitaldan I updated the wiki to explain all the optional parameters better. If you have any more edits, please feel free.

digitaldan commented:
@computergeek1507 the explanation on the wiki looks great, thanks for updating it!

digitaldan commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBinding.java:
```diff
@@ -0,0 +1,389 @@
+										break;
+									}
+								} else if (PercentType.class == type) {
+									if (garageopener.getStatus() != GarageDoorStatus.UNKNOWN) {
```
I don't think this is right, it should be
```
} else if (PercentType.class == type) {
									if(garageopener.getStatus() == GarageDoorStatus.OPEN){
										newState = new PercentType(0);
									} else if (garageopener.getStatus() == GarageDoorStatus.CLOSED){
										newState = new PercentType(100);
									} else if (garageopener.getStatus() != GarageDoorStatus.UNKNOWN) {
										//if we are not open, closed or unknow, we are partially opened
										newState = new PercentType(50);
										break;
									}
								} 
```

digitaldan replied:
That didn't format as I thought it would, let me do a quick pr ;-)

computergeek1507 replied:
like this:
if (PercentType.class == type) {
									if (garageopener.getStatus() == GarageDoorStatus.CLOSED) {
										newState = PercentType.HUNDRED;
										break;
									} else if (garageopener.getStatus() == GarageDoorStatus.OPEN) {
										newState = PercentType.ZERO;
										break;
									}
									else if (garageopener.getStatus() != GarageDoorStatus.UNKNOWN) {
										newState = new PercentType(50);
										break;
									}
								}

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBinding.java:
```diff
@@ -0,0 +1,400 @@
+								} else if (PercentType.class == type) {
+									switch (garageopener.getStatus()) {
+									case OPEN:
+										newState = new PercentType(0);
```
`PercentType.ZERO` would mean one less object creation.

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBinding.java:
```diff
@@ -0,0 +1,400 @@
+										newState = new PercentType(0);
+										break;
+									case CLOSED:
+										newState = new PercentType(100);
```
`PercentType.HUNDRED` would mean one less object creation.

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqBinding.java:
```diff
@@ -0,0 +1,400 @@
+	}
+
+	/**
+	 * @{inheritDoc
```
Use `{@inheritDoc}`

watou commented on bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqData.java:
```diff
@@ -0,0 +1,229 @@
+		logger.debug("Retreiveing door data");
+		String url = String.format(
+				"%s/api/v4/userdevicedetails/get?appId=%s&SecurityToken=%s",
+				WEBSITE, appId, getSecurityToken());
```
Shouldn't `appId` and `SecurityToken` values be URL-encoded (or any URL parameters that could potentially contain illegal characters elsewhere in this source file)?  I see the default app ID has `/` and `+` in them, for example.  The appId might be only accidentally working at present.

watou commented:
I think we're just about there.  Thank you for all your efforts in contributing this new binding.  Could you take a final sweep through for any remaining small issues, make sure the Eclipse formatter has been applied to all Java files, rebase against master to resolve conflicts, do final testing, and squash into one commit?

teichsta commented:
> I think we're just about there. Thank you for all your efforts in contributing this new binding. Could you take a final sweep through for any remaining small issues, make sure the Eclipse formatter has been applied to all Java files, rebase against master to resolve conflicts, do final testing, and squash into one commit?

+1

computergeek1507 commented:
@watou @teichsta I used the the Eclipse formatter and rebase against master. I have been running the binding on my main openHAB 1.7.1 server with zero issues for the last couple weeks and I have also tested it with the openHAB 2.0 snapshot. I squash my commits and have made a wiki here:
https://github.com/computergeek1507/openhab/wiki/Chamberlain-MyQ-Binding

If there is anything else I need to do, please let me know.

Thanks!

teichsta commented:
great thanks! @watou would you like to go ahead an merge?

watou commented:
All merged.  Could @computergeek1507 or @digitaldan create the wiki page in the openhab/openhab/wiki, hook it into the sidebar, and add a note in the wiki page that the binding is available as of openHAB 1.8?  Thanks for all the hard work and persistence through the review process!

teichsta commented:
> Thanks for all the hard work and persistence through the review process!

+1 great work, guys!

computergeek1507 commented:
@watou Wiki Page is created 
https://github.com/openhab/openhab/wiki/Chamberlain-MyQ-Binding

watou commented:
Very nicely done!  Thank you.
